include/envoy/api/os_sys_calls.h

```diff
@@ -185,6 +185,14 @@ class OsSysCalls {
    * @see man TCP_INFO. Get the tcp info for the socket.
    */
   virtual SysCallBoolResult socketTcpInfo(os_fd_t sockfd, EnvoyTcpInfo* tcp_info) PURE;
+
+  /**
+   * @see MSDN WSAIoctl. Controls the mode of a socket.
+   */
+  virtual SysCallIntResult win32Ioctl(os_fd_t sockfd, unsigned long control_code, void* in_buffer,
+                                      unsigned long in_buffer_len, void* out_buffer,
+                                      unsigned long out_buffer_len,
+                                      unsigned long* bytes_returned) PURE;
```

Comment:
Apologies if this was already discussed, but it seems odd to me to add this API at the global level as it is windows specific (yes I know the rest of the APIs are POSIX specific but there is some assumption that all OS support POSIX in some form). Since the syscalls are accessed by singleton, wdyt about adding a Win32 specific syscalls class that is only compiled on win32, then checked to be only called within win32 code? Since this gets called through from other APIs like socket, I'm wondering if additionally there could be a set of win32 specific free functions that know how to do various things with these special IOCTLs vs. having NOP everywhere else? The alternative would be to define some type of "extra" generic IOCTL and have it be feature based vs. OS based, but that seems like more work?

Reply:
Is the approach that you are describing to remove `win32Ioctl` from the `socket`/`io_handle` interface? And invoke this functionality through a  new  `win32OsSysCallsSingleton`?

--- 
Another approach I considered was to change the API signature of `ioctl` in `os_sys_calls.h` to mirror the Windows function signature. 

i.e.

instead of 
```
  /**
   * @see ioctl (man 2 ioctl)
   */
  virtual SysCallIntResult ioctl(os_fd_t sockfd, unsigned long int request, void* argp) PURE;
```

to have:

```
  /**
   * @see MSDN WSAIoctl. Controls the mode of a socket.
   */
  virtual SysCallIntResult win32Ioctl(os_fd_t sockfd, unsigned long control_code, void* in_buffer,
                                      unsigned long in_buffer_len, void* out_buffer,
                                      unsigned long out_buffer_len,
                                      unsigned long* bytes_returned) PURE;
```

and in POSIX the last parameters would be unused (`in_buffer` would correspond to `argp`). There are few use cases for `ioctl` in the codebase (I couldn't find any caller outside of tests).

Reply:
I was suggesting the former, but I think the latter is fine also and simpler with appropriate comments. WDYT?

Reply:
I prefer the later, because in the former we would need to have access to the `int fd` to invoke the system calls which is not always easy since we abstract it away from the caller in the `io_socket_handle_impl`.

Reply:
fixed in https://github.com/envoyproxy/envoy/pull/14753/commits/f7698962ac3e21e673937b4e3723f171b87ae8ea